File: Leetcode/Array&String/169_majority-element/main.py
from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        def count_in_range(target, low, high):
            return sum(1 for i in range(low, high + 1) if nums[i] == target)

        def majority_element_rec(low, high):
            if low == high:
                return nums[low]
            mid = (low + high) // 2
            left_majority = majority_element_rec(low, mid)
            right_majority = majority_element_rec(mid + 1, high)

            left_count = count_in_range(left_majority, low, high)
            right_count = count_in_range(right_majority, low, high)
            return left_majority if left_count > right_count else right_majority

        return majority_element_rec(0, len(nums) - 1)


if __name__ == '__main__':
    nums = [2, 2, 1, 1, 1, 2, 2]
    res = Solution().majorityElement(nums)
    print(res)

File: Leetcode/Array&String/169_majority-element/main.cpp
// 169. 多数元素
// Created by bob on 2022/7/26.
//

#include<vector>
#include<unordered_map>

using namespace std;

/**
 * 方法二：分治
 */
class Solution {
private:
    int countInRange(vector<int>& nums, int target, int low, int high) {
        int count = 0;
        for (int i = low; i <= high; ++i) {
            if (nums[i] == target)
                count++;
        }
        return count;
    }
    
    int majorityElementRec(vector<int>& nums, int low, int high) {
        if (low == high) {
            return nums[low];
        }
        int mid = (low + high) / 2;
        int left_majority = majorityElementRec(nums, low, mid);
        int right_majority = majorityElementRec(nums, mid + 1, high);
        if (countInRange(nums, left_majority, low, high) > (high - low + 1) / 2) {
            return left_majority;
        }
        if (countInRange(nums, right_majority, low, high) > (high - low + 1) / 2) {
            return right_majority;
        }
        return -1;
    }
public:
    int majorityElement(vector<int>& nums) {
        return majorityElementRec(nums, 0, nums.size() - 1);
    }
};

/**
 * 方法一：hash表
 */
class Solution2 {
public:
    int majorityElement(vector<int>& nums) {
        unordered_map<int, int> counts;
        int majority = 0, cnt = 0;
        for (int num: nums) {
            counts[num]++;
            if (counts[num] > cnt) {
                majority = num;
                cnt = counts[num];
            }
        }
        return majority;
    }
};


int main() {
    vector<int> nums = {2, 2, 1, 1, 1, 2, 2};
    int res = Solution().majorityElement(nums);
    int res2 = Solution2().majorityElement(nums);
    printf("%d", res2);
    return 0;
}